multicluster/service-mirror/cluster_watcher.go

```diff
@@ -815,3 +913,296 @@ func (rcsw *RemoteClusterServiceWatcher) createOrUpdateEndpoints(ctx context.Con
+		})
+	}
+
+	headlessMirrorServiceName := rcsw.mirroredResourceName(exportedService.Name)
+	headlessMirrorEndpoints := &corev1.Endpoints{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      headlessMirrorServiceName,
+			Namespace: exportedService.Namespace,
+			Labels: map[string]string{
+				consts.MirroredResourceLabel:  "true",
+				consts.RemoteClusterNameLabel: rcsw.link.TargetClusterName,
+			},
+			Annotations: map[string]string{
+				consts.RemoteServiceFqName: fmt.Sprintf("%s.%s.svc.%s", exportedService.Name, exportedService.Namespace, rcsw.link.TargetClusterDomain),
+			},
+		},
+		Subsets: subsetsToCreate,
+	}
+
+	if rcsw.link.GatewayIdentity != "" {
+		headlessMirrorEndpoints.Annotations[consts.RemoteGatewayIdentity] = rcsw.link.GatewayIdentity
+	}
+
+	rcsw.log.Infof("Creating a new Headless Mirror endpoints object for Headless Mirror %s/%s", headlessMirrorServiceName, exportedService.Namespace)
+	if _, err := rcsw.localAPIClient.Client.CoreV1().Endpoints(exportedService.Namespace).Create(ctx, headlessMirrorEndpoints, metav1.CreateOptions{}); err != nil {
+		// we clean up after ourselves
+		rcsw.localAPIClient.Client.CoreV1().Services(exportedService.Namespace).Delete(ctx, headlessMirrorServiceName, metav1.DeleteOptions{})
+		// and retry
+		return RetryableError{[]error{err}}
+	}
+
+	return nil
+}
+
+// createEndpointMirrorService creates a new Endpoint Mirror service and its
+// corresponding endpoints object. It returns the newly created Endpoint Mirror
+// service object. When a headless service is exported, we create a Headless
+// Mirror service in the source cluster and then for each hostname in the
+// exported service's endpoints object, we also create an Endpoint Mirror
+// service (and its corresponding endpoints object).
+func (rcsw *RemoteClusterServiceWatcher) createEndpointMirrorService(ctx context.Context, endpointHostname, resourceVersion, endpointMirrorName string, exportedService *corev1.Service) (*corev1.Service, error) {
+	gatewayAddresses, err := rcsw.resolveGatewayAddress()
+	if err != nil {
+		return nil, err
+	}
+
+	endpointMirrorAnnotations := map[string]string{
+		// needed to detect real changes
+		consts.RemoteResourceVersionAnnotation: resourceVersion,
+		consts.RemoteServiceFqName:             fmt.Sprintf("%s.%s.%s.svc.%s", endpointHostname, exportedService.Name, exportedService.Namespace, rcsw.link.TargetClusterDomain),
+	}
+
+	endpointMirrorLabels := rcsw.getMirroredServiceLabels()
+	mirrorServiceName := rcsw.mirroredResourceName(exportedService.Name)
+	endpointMirrorLabels[consts.MirroredHeadlessSvcNameLabel] = mirrorServiceName
+
+	// Create service spec, clusterIP
+	endpointMirrorService := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        endpointMirrorName,
+			Namespace:   exportedService.Namespace,
+			Annotations: endpointMirrorAnnotations,
+			Labels:      endpointMirrorLabels,
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: remapRemoteServicePorts(exportedService.Spec.Ports),
```

Comment:
So it looks like we need to get ports a different way here. From what I can tell, it's fine for a headless service to not have `spec.ports`; this is the case for my nginx service (posted above in a comment). However, when we created a mirror cluster IP service here it _does_ need `spec.ports`, but we have none to copy from the exported service because it is headless.

In the case of my nginx example the ports don't seem to be easy to get. It exposes port `80`, but I don't actually see that in the endpoints object:

```
❯ kubectl --context k3d-x get endpoints nginx-svc -o yaml
apiVersion: v1
kind: Endpoints
metadata:
  creationTimestamp: "2021-06-29T22:17:41Z"
  labels:
    mirror.linkerd.io/exported: "true"
    service.kubernetes.io/headless: ""
  managedFields:
  ...
subsets:
- addresses:
  - hostname: nginx-ss-0
    ip: 10.42.0.19
    nodeName: k3d-x-server-0
    targetRef:
      kind: Pod
      name: nginx-ss-0
      namespace: default
      resourceVersion: "2852"
      uid: 972afe4e-86ac-4ebc-bb13-04306656db68
  - hostname: nginx-ss-1
    ip: 10.42.0.20
    nodeName: k3d-x-server-0
    targetRef:
      kind: Pod
      name: nginx-ss-1
      namespace: default
      resourceVersion: "2915"
      uid: 6c3fee31-7f67-4990-88b1-5c32ab6bc6cb
  - hostname: nginx-ss-2
    ip: 10.42.0.22
    nodeName: k3d-x-server-0
    targetRef:
      kind: Pod
      name: nginx-ss-2
      namespace: default
      resourceVersion: "2943"
      uid: 3488e524-0491-418b-8b29-8e751ccf03ca
```

or even the pods:
```
❯ kubectl --context k3d-x get pods nginx-ss-0 -o yaml
apiVersion: v1
kind: Pod
metadata:
  ...
spec:
  containers:
  - name: linkerd-proxy
    ...
  - image: nginx
    imagePullPolicy: Always
    name: nginx
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-22hf5
      readOnly: true
    ...
status:
  ...
```

It's possible this is a special case because it only exposes port `80` so that is just the default?

Reply:
Ugh, this is news to me. I looked into it and can confirm headless services are allowed to be created without a `Ports` field. I think the rationale here is that they may be created since they serve to set up DNS entries which makes the port optional. It turns out, when the clusterIP is set to None or to an external address, Endpoints are still created. [1](https://github.com/kubernetes/kubernetes/issues/49686) [2](https://github.com/kubernetes/kubernetes/pull/47250). Unfortunately, this is not documented in the official k8s docs (or at least no docs that I know of). We can't really get around this since our endpoint mirrors will require a port to be created.

Not sure how to proceed here. My first instinct would be to skip exported services with no ports. This is definitely the easiest way. We could look at the backing pods but I'm not sure what to do when the pod objects don't have the ports in the spec either. wdyt @kleimkuhler.

Reply:
Done. Added a recorder to the `RemoteClusterWatcher` structure that we instantiate whenever we instantiate the watcher. The recorder will send events to the `remote` cluster and associate them with the service object that cannot be exported/mirrored:
```
37s         Normal    ServiceMirroringSkipped   service/nginx-svc        Skipped mirroring service: object spec has no exposed ports
```